core/shared/src/main/scala-3/zio/HasNoScopeCompanionVersionSpecific.scala

```diff
@@ -0,0 +1,39 @@
+package zio
+
+import scala.quoted._
+
+private[zio] abstract class HasNoScopeCompanionVersionSpecific {
+
+  val instance: HasNoScope[Any]
+
+  final transparent inline given hasNoScope[R]: HasNoScope[R] =
+    ${ HasNoScopeMacro.noScope[R] }
+
+}
+
+private[zio] object HasNoScopeMacro {
+
+  def noScope[R: Type](using Quotes): Expr[HasNoScope[R]] = {
```

Comment:
Just curious: have you tried `NotGiven[R <: Scope]`?
https://www.scala-lang.org/api/3.4.0/scala/util/NotGiven.html

Reply:
this does not solve this use case
```scala
 test("generic with R") {
        typeCheck(
          // language=Scala
          """
               def genericNoImplicitWithR[R]: ZIO[Int & R, Nothing, Unit] =
                noScope[Int & R] // Here, NotGiven[R <: Scope] is always available even if R does contain Scope
            """
        ).map(e =>
          assertTrue(
            e == Left("Can not prove that R does not contain a zio.Scope. Please add a context bound R: HasNoScope.")
          )
        )
      }
```
With the macro, you will get an error asking you to add a context bound.
With NotGiven, a generic `R` can never be proven to contain a `Scope` so the NotGiven is always found.